python/ray/serve/_private/replica.py

```diff
@@ -494,95 +515,74 @@ def callable_method_filter(attr):
+            f"Started executing request {request_metadata.request_id}",
             extra={"log_to_stderr": False},
         )
 
-        args, kwargs = parse_request_item(request_item)
+        if request_metadata.is_http_request:
+            assert len(request_args) == 3
+            scope, receive, send = request_args
 
-        # Check if the callable is our ASGI wrapper (i.e., the user used
-        # `@serve.ingress`).
-        callable_is_asgi_wrapper = isinstance(self.callable, ASGIAppReplicaWrapper)
-        if asgi_sender is not None and callable_is_asgi_wrapper:
-            kwargs["asgi_sender"] = asgi_sender
+            if isinstance(self.callable, ASGIAppReplicaWrapper):
+                request_args = (scope, receive, send)
+            else:
+                request_args = (Request(scope, receive, send),)
 
         method_to_call = None
         success = True
         try:
-            runner_method = self.get_runner_method(request_item)
+            runner_method = self.get_runner_method(request_metadata)
             method_to_call = sync_to_async(runner_method)
             result = None
-            if len(inspect.signature(runner_method).parameters) > 0:
-                result = await method_to_call(*args, **kwargs)
-            else:
-                # When access via http http_arg_is_pickled with no args:
-                # args = (<starlette.requests.Request object at 0x7fe900694cc0>,)
-                # When access via python with no args:
-                # args = ()
-                if len(args) == 1 and isinstance(args[0], starlette.requests.Request):
-                    # The method doesn't take in anything, including the request
-                    # information, so we pass nothing into it
-                    result = await method_to_call()
-                else:
-                    # Will throw due to signature mismatch if user attempts to
-                    # call with non-empty args
-                    result = await method_to_call(*args, **kwargs)
 
-            # Streaming HTTP codepath: always send response over ASGI interface.
-            if asgi_sender is not None:
+            # Edge case to support empty HTTP handlers: don't pass the Request
+            # argument if the callable has no parameters.
+            if (
+                request_metadata.is_http_request
+                and len(inspect.signature(runner_method).parameters) == 0
+            ):
+                request_args, request_kwargs = tuple(), {}
+
+            result = await method_to_call(*request_args, **request_kwargs)
+            if request_metadata.is_http_request:
                 # For the FastAPI codepath, the response has already been sent over the
                 # ASGI interace and result should always be `None`.
-                if callable_is_asgi_wrapper:
+                if isinstance(self.callable, ASGIAppReplicaWrapper):
                     assert result is None
```

Comment:
Non-blocker: Would be nice if we can also log a message explain this assertion. But knowing this was already like that before and can be improved in a followup.